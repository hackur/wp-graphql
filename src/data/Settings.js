import { toGlobalId } from 'graphql-relay';
import Dataloader from 'dataloader';
import fetchData from 'data/utils';

// Dataloader expects IDs that can be read by the REST API

// there is no batch mechanism on this endpoint
const path = process.env.WP_SETTINGS_ENDPOINT || 'graphql/v1/settings';
const settingsLoader = new Dataloader(settingsPaths =>
  Promise.all(
    settingsPaths.map(settingsPath =>
      fetchData(settingsPath).then(({ data: { body } }) => body)
    )
  )
);

class Settings {
  getID() {
    return toGlobalId(this.constructor.name, 'me');
  }

  static getEndpoint() {
    return path;
  }

  static async load() {
    const settings = await settingsLoader.load(path);
    return settings ? Object.assign(new Settings(), settings) : null;
  }
}

export default Settings;
