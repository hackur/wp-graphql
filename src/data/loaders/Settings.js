import Dataloader from 'dataloader';
import fetchData from 'data/utils';
import Settings from 'data/Settings';

// there is no batch mechanism on this endpoint
const endpoint = Settings.getEndpoint();

// Dataloader expects IDs that can be read by the REST API

export default function getSettingsLoaders() {
  const settingsLoader = new Dataloader(settingsPaths =>
    Promise.all(
      settingsPaths.map(settingsPath =>
        fetchData(settingsPath).then(({ data: { body } }) => body)
      )
    )
  );

  return {
    load: async () => {
      const settings = await settingsLoader.load(endpoint);
      return settings ? Object.assign(new Settings(), settings) : null;
    },
  };
}
