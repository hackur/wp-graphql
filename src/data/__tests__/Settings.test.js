import Settings from 'data/Settings';

describe('Test Settings data access', () => {
  test('Get endpoint', () => {
    expect(Settings.getEndpoint()).toMatchSnapshot();
  });

  test('Load settings', async () => {
    const settings = await Settings.load();
    expect(settings.getID()).toMatchSnapshot();
    expect(settings).toMatchSnapshot();
  });
});
