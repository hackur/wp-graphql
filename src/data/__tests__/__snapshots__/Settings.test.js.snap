// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Settings data access Get endpoint 1`] = `"graphql/v1/settings"`;

exports[`Test Settings data access Load settings 1`] = `"U2V0dGluZ3M6bWU="`;

exports[`Test Settings data access Load settings 2`] = `
Settings {
  "slug": "graphql/v1/settings",
}
`;
